examples/qualcomm/llama2/model/static_llama.py

```diff
@@ -0,0 +1,253 @@
+        kv_mask: torch.Tensor,
+    ) -> Tuple[torch.Tensor]:
+        h, k_cache, v_cache = self.attention(
+            hidden_states=self.attention_norm(x),
+            freqs_cos=freqs_cos,
+            freqs_sin=freqs_sin,
+            atten_mask=atten_mask,
+            k_cache=k_cache,
+            v_cache=v_cache,
+            kv_mask=kv_mask,
+        )
+        h = x + h
+        output = h + self.feed_forward(self.ffn_norm(h))
+        return output, k_cache, v_cache
+
+
+class LlamaModel(nn.Module):
+    def __init__(self, config: ModelArgs):
+        super().__init__()
+        self.dim = config.dim
+        self.max_batch_size = config.max_batch_size
+        self.max_seq_len = config.max_seq_len
+        self.n_heads = config.n_heads
+        self.n_layers = config.n_layers
+        self.vocab_size = config.vocab_size
+
+        self.layers = nn.ModuleList(
+            [LlamaDecoderLayer(config) for _ in range(config.n_layers)]
+        )
+        self.norm = RMSNorm(config.dim, eps=config.norm_eps)
+        self.output = nn.Linear(config.dim, config.vocab_size, bias=False)
+        self.tok_embeddings = nn.Embedding(config.vocab_size, config.dim)
+        freqs_cos, freqs_sin = precompute_freqs_cis(
+            config.dim // config.n_heads,
+            config.max_seq_len,
+            config.rope_freq_base,
+        )
+        self.register_buffer("freqs_cos", freqs_cos, persistent=False)
+        self.register_buffer("freqs_sin", freqs_sin, persistent=False)
+        self.register_buffer("mask", torch.ones(self.dim), persistent=False)
+        self.register_buffer("unmask", torch.zeros(self.dim), persistent=False)
+
+    def forward(
+        self,
+        tokens: torch.Tensor,
+        input_pos: torch.Tensor,
+        atten_mask: torch.Tensor,
+        k_cache: torch.Tensor,
+        v_cache: torch.Tensor,
+        kv_mask: torch.Tensor,
+    ) -> Tuple[torch.Tensor, List[torch.Tensor], List[torch.Tensor]]:
+        output_k_cache = []
+        output_v_cache = []
+        # following tensors should be invariant across batches
+        freqs_cos = self.freqs_cos[input_pos][0]
+        freqs_sin = self.freqs_sin[input_pos][0]
+        atten_mask = atten_mask[input_pos][0]
+
+        hidden_states = self.tok_embeddings(tokens)
+        for ind, decoder_layer in enumerate(self.layers):
+            k = k_cache[:, ind, :, :]
+            v = v_cache[:, ind, :, :]
+            hidden_states, k, v = decoder_layer(
+                hidden_states,
+                freqs_cos=freqs_cos,
+                freqs_sin=freqs_sin,
+                atten_mask=atten_mask,
+                k_cache=k,
+                v_cache=v,
+                kv_mask=kv_mask,
+            )
+            output_k_cache.append(
+                k.view(self.max_batch_size, self.max_seq_len, self.dim)
+            )
+            output_v_cache.append(
+                v.view(self.max_batch_size, self.max_seq_len, self.dim)
+            )
```

Comment:
Aren't we having a dynamic shape for `output_k_cache` here?

Reply:
No, we're concatenating kv_cache calculated from all attention layers:

- Input shape of k_cache: (max_batch_size, n_layers, max_seq_len, embedding_dim)
- Slice for every attention layer: (max_batch_size, max_seq_len, embedding_dim)

We could concatenate them back without runtime shape changing.

Reply:
hmm probably need to check the graph, I just feel like `output_v_cache` shape is changing when iterating the layer, because we keep appending it

Reply:
The list append method will be changed after `torch.export`. The output kv_caches from each attention layer are connected directly to final `concat` operator in `LlamaModel.forward`.